lib/wasi/src/state/builder.rs

```diff
@@ -359,29 +385,35 @@ impl WasiStateBuilder {
             }
         }
 
-        // self.preopens are checked in [`PreopenDirBuilder::build`]
+        // using `take` here is a bit of a code smell
```

Comment:
Yeah, totally agree. If the `build` method was taking ownership of `self` rather than a mutable reference, that would not be an issue because `self` wouldn't be any longer usable. But in this case, it changes the behavior of the object. I don't like it at all :-). It must be changed!

Reply:
Resolution: make `build` consume `self` (so we don't need to wrap in an `Arc` or other paths)

Reply:
It's not possible because it will break existing code, like:

```
let mut builder = WasiStateBuilder::new();
builder.stdin(…);
builder.set_fs(…);
```

it won't compile anymore because `stdin` will consume `self`.

Except if we introduce a BC break here, it's not possible to solve this problem. I've decided to write what's happening in the documentation when calling `build` or `finalize` multiple times. I think it's the best trade-off for the moment.

See https://github.com/wasmerio/wasmer/pull/2491/commits/4eaa864ae8b08daa3dd97f97c633ad0e6a6b6583.